src/components/ImagePreviewer.tsx

```diff
@@ -0,0 +1,253 @@
+    position: absolute;
+    border-width: 1px;
+`;
+
+const DeleteButton = styled.div`
+    position: absolute;
+    padding: 4px;
+    background: rgba(255, 0, 0, 0.3);
+    cursor: pointer;
+`;
+
+interface ImagePreviewerProps {
+    ownedRects: Rect[];
+    setOwnedRects: (rects: Rect[]) => void;
+    imageData: string;
+}
+
+const genId = () => Date.now().toString(16);
+
+const ImagePreviewer = (props: ImagePreviewerProps) => {
+    const imagePreviewerRef = useRef<HTMLDivElement>(null);
+    const { ownedRects, setOwnedRects, imageData } = props;
+
+    const handleImagePreviewerMouseDown = (e: MouseEvent<HTMLDivElement>) => {
+        if (!imagePreviewerRef.current) {
+            return alert("image previewer is not ready");
+        }
+
+        const bounding = imagePreviewerRef.current.getBoundingClientRect();
+
+        const relativeX = e.clientX - bounding.x;
+        const relativeY = e.clientY - bounding.y;
+
+        if (ownedRects.some((rect) => rect.isMoving)) {
+            return setOwnedRects(
+                ownedRects.map((rect) => {
+                    rect.isMoving = false;
+                    return rect;
+                })
+            );
+        }
+
+        if (
+            ownedRects.length &&
+            !ownedRects[ownedRects.length - 1].isCompleted
+        ) {
+            return setOwnedRects(
+                ownedRects.map((v, i, arr) => {
+                    if (i !== arr.length - 1) {
+                        return v;
+                    }
+                    const nextRect: Rect = {
+                        ...v,
+                        width: relativeX - v.x,
+                        height: relativeY - v.y,
+                        isCompleted: true,
+                    };
+
+                    if (nextRect.width < 0) {
+                        nextRect.x = nextRect.x + nextRect.width;
+                        nextRect.width = Math.abs(nextRect.width);
+                    }
+
+                    if (nextRect.height < 0) {
+                        nextRect.y = nextRect.y + nextRect.height;
+                        nextRect.height = Math.abs(nextRect.height);
+                    }
+
+                    return nextRect;
+                })
+            );
+        }
+
+        let found = false;
+        const nextRects = Array.from(ownedRects)
+            .reverse()
+            .map((rect) => {
+                if (found) {
+                    return rect;
+                }
+                if (
+                    (relativeX - rect.x) * (relativeX - rect.x - rect.width) <
+                        0 &&
+                    (relativeY - rect.y) * (relativeY - rect.y - rect.height) <
+                        0
+                ) {
+                    found = true;
+                    return {
+                        ...rect,
+                        isMoving: true,
+                    };
+                }
+                return rect;
+            })
+            .reverse();
```

Comment:
1. I would like to select the latest created rectangle and don't want to change the orders of the rectangles. It could also be done by finding the last match and map it again.
2. It's computing if a point p with the coordinate `(relativeX, relativeY)` is inside the rectangle which diagonal is `p1(rect.x, rect.y)` and `p2(rect.x + rect.width, rect.y + rect.height)`. The rectangle is not skewed and not rotated so we could say p is inside the rectangle if p.x is between p1.x and p2.x and p.y is between p1.y and p2.y.
```
// relativeX is between `rect.x` and `rect.x + rect.width`
(relativeX - rect.x) * (relativeX - rect.x - rect.width) < 0
// relativeY is between `rect.y` and `rect.y + rect.height`
(relativeY - rect.y) * (relativeY - rect.y - rect.height) < 0